Refactor list-style-type conversion

vitorroriz opened the pull request:
#### 2fa477e5ad2efadf91a266ffc92ca87e2885fb72
<pre>
Refactor list-style-type conversion
<a href="https://bugs.webkit.org/show_bug.cgi?id=253819">https://bugs.webkit.org/show_bug.cgi?id=253819</a>
rdar://106640014

list-style-type is currently used for setting a list style type and a list style
string value. Currently, at StyleBuilderCustom we always sets both values together.
This refactoring will introduce a new type to explicitly associate these values, ListStyle.
ListStyle has a a type and a name (string value).
This is a preparation for @counter-style, since list-style-type&apos;s associated assigned with a custom-ident
will always have CustomCounterStyle as ListStyleType (ListStyle.type) and
will rely on the ListStyle.name to differentiate between different counter-styles.

Reviewed by Antti Koivisto.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilityList.cpp:
(WebCore::AccessibilityList::determineAccessibilityRole):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/editing/Editor.cpp:
(WebCore::editableTextListsAtPositionInDescendingOrder):
* Source/WebCore/editing/FontAttributes.h:
* Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
(WebCore::cocoaTextListMarkerName):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::effectiveListMarkerType):
(WebCore::listMarkerSuffix):
(WebCore::listMarkerText):
(WebCore::RenderListMarker::styleDidChange):
(WebCore::RenderListMarker::textRun const):
(WebCore::RenderListMarker::paint):
(WebCore::RenderListMarker::updateContent):
(WebCore::RenderListMarker::computePreferredLogicalWidths):
(WebCore::RenderListMarker::updateMargins):
(WebCore::RenderListMarker::relativeMarkerRect):
(WebCore::RenderListMarker::counterStyle const):
* Source/WebCore/rendering/RenderListMarker.h:
* Source/WebCore/rendering/style/CounterContent.h:
(WebCore::CounterContent::CounterContent):
(WebCore::CounterContent::listStyle const):
* Source/WebCore/rendering/style/ListStyleType.cpp: Added.
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/ListStyleType.h: Added.
(WebCore::ListStyleType::operator== const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::rareInheritedDataChangeRequiresLayout):
(WebCore::RenderStyle::changeRequiresLayout const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::listStyleType const):
(WebCore::RenderStyle::setListStyleType):
(WebCore::RenderStyle::initialListStyleType):
(WebCore::RenderStyle::InheritedFlags::operator== const):
(WebCore::RenderStyle::listStyleStringValue const): Deleted.
(WebCore::RenderStyle::setListStyleStringValue): Deleted.
(WebCore::RenderStyle::initialListStyleStringValue): Deleted.
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
(WebCore::RenderTreeBuilder::List::updateItemMarker):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertListStyleType):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):
(WebCore::Style::BuilderCustom::applyInheritListStyleType): Deleted.
(WebCore::Style::BuilderCustom::applyInitialListStyleType): Deleted.
(WebCore::Style::BuilderCustom::applyValueListStyleType): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/261651@main">https://commits.webkit.org/261651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8070be674af22d21c74d6a4bd6f93eb469ebbd0f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21600 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1104 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4232 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/121035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116519 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12747 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/5379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118220 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100231 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105509 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13950 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/814 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14644 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/19983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52818 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8124 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16462 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WebCore/rendering/style/ListStyle.h:
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "RenderStyleConstants.h"
+#include <utility>
```
I don't need this header anymore. I'll get rid of it.

anttijk replied:
> I don't need this header anymore. I'll get rid of it.

Ok

anttijk commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 51;
+	objectVersion = 54;
```
huh?

vitorroriz replied:
hmmmm no idea what my Xcode did

anttijk commented on Source/WebCore/rendering/style/ListStyle.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace WebCore {
+struct ListStyle {
+    ListStyleType type { ListStyleType::None };
```
The natural name for the struct is `ListStyleType` since it is the value of the property `list-style-type` but that is annoyingly taken by the enum. Maybe the enum could be in the struct scope as `ListStyleType::Type`?

Failing that `ListStyleTypeValue`? Or you could rename enum to something? I don't have great ideas though.

nt1m replied:
+1 to `ListStyleType::Type`

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -3054,4 +3054,9 @@ UserSelect RenderStyle::effectiveUserSelect() const
+{
+    m_inheritedFlags.listStyleType = static_cast<unsigned>(listStyle.type);
+    setListStyleStringValue(listStyle.name);
+}
```
This is rather awkward. You should probably move the whole ListStyleType value to RareInheritedData instead of having the bit in RenderStyle and the text value in rare data. I'd think it is reasonably rare.

anttijk commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -3998,7 +3998,7 @@
             ],
             "codegen-properties": {
-                "custom": "All",
+                "converter": "ListStyleType",
```
Cool

nt1m commented on Source/WebCore/rendering/style/ListStyle.h:
```diff
@@ -0,0 +1,45 @@
+namespace WebCore {
+struct ListStyle {
+    ListStyleType type { ListStyleType::None };
+    AtomString name;
```
identifier ? I don't like `name` because in the string case, it's not really representing a name, but rather the string to be displayed.

nt1m commented on Source/WebCore/rendering/style/ListStyle.h:
```diff
@@ -0,0 +1,45 @@
+inline WTF::TextStream& operator<<(WTF::TextStream& ts, ListStyle listStyle)
+{
+    ts << "list-style-type: "_s << listStyle.type << " list-style-string-value:" << listStyle.name;
+    return ts;
```
Can we reflect the exact CSS presentation?

nt1m commented on Source/WebCore/rendering/style/ListStyleType.cpp:
```diff
@@ -0,0 +1,48 @@
+{
+    if (listStyle.type == ListStyleType::Type::CustomCounterStyle)
+        ts << listStyle.identifier;
+    else
```
ListStyleType::Type::String should log the identifier with quotes around it.

vitorroriz replied:
ops, good catch, thanks!

vitorroriz committed:
Refactor list-style-type conversion
https://bugs.webkit.org/show_bug.cgi?id=253819
rdar://106640014

list-style-type is currently used for setting a list style type and a list style
string value. Currently, at StyleBuilderCustom we always sets both values together.
This refactoring will introduce a new type to explicitly associate these values, ListStyle.
ListStyle has a a type and a name (string value).
This is a preparation for @counter-style, since list-style-type's associated assigned with a custom-ident
will always have CustomCounterStyle as ListStyleType (ListStyle.type) and
will rely on the ListStyle.name to differentiate between different counter-styles.

Reviewed by Antti Koivisto.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilityList.cpp:
(WebCore::AccessibilityList::determineAccessibilityRole):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/editing/Editor.cpp:
(WebCore::editableTextListsAtPositionInDescendingOrder):
* Source/WebCore/editing/FontAttributes.h:
* Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
(WebCore::cocoaTextListMarkerName):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::effectiveListMarkerType):
(WebCore::listMarkerSuffix):
(WebCore::listMarkerText):
(WebCore::RenderListMarker::styleDidChange):
(WebCore::RenderListMarker::textRun const):
(WebCore::RenderListMarker::paint):
(WebCore::RenderListMarker::updateContent):
(WebCore::RenderListMarker::computePreferredLogicalWidths):
(WebCore::RenderListMarker::updateMargins):
(WebCore::RenderListMarker::relativeMarkerRect):
(WebCore::RenderListMarker::counterStyle const):
* Source/WebCore/rendering/RenderListMarker.h:
* Source/WebCore/rendering/style/CounterContent.h:
(WebCore::CounterContent::CounterContent):
(WebCore::CounterContent::listStyle const):
* Source/WebCore/rendering/style/ListStyleType.cpp: Added.
(WebCore::operator<<):
* Source/WebCore/rendering/style/ListStyleType.h: Added.
(WebCore::ListStyleType::operator== const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::rareInheritedDataChangeRequiresLayout):
(WebCore::RenderStyle::changeRequiresLayout const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::listStyleType const):
(WebCore::RenderStyle::setListStyleType):
(WebCore::RenderStyle::initialListStyleType):
(WebCore::RenderStyle::InheritedFlags::operator== const):
(WebCore::RenderStyle::listStyleStringValue const): Deleted.
(WebCore::RenderStyle::setListStyleStringValue): Deleted.
(WebCore::RenderStyle::initialListStyleStringValue): Deleted.
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeBuilderList.cpp:
(WebCore::RenderTreeBuilder::List::updateItemMarker):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertListStyleType):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueContent):
(WebCore::Style::BuilderCustom::applyInheritListStyleType): Deleted.
(WebCore::Style::BuilderCustom::applyInitialListStyleType): Deleted.
(WebCore::Style::BuilderCustom::applyValueListStyleType): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/261651@main
